### DISCUSSION TITLE
How to pass processArgument using driver.executeScript("mobile: installApp", args)

### DISCUSSION THREAD

**Discussion by deepakarorawins** [appium/appium#19869] [Q&A]
@mykola-mokhnach @imurchie 

https://github.com/appium/appium/issues/9659

As mentioned in above link we can set processArgument using capability, is there any way we can pass processArguments using executeScript commands?

Why?: We need to pass processArgument which is working fine when driver initialized first. then on executeScript it loose that process argument.

getDriver().executeScript("mobile: removeApp", args);
getDriver().executeScript("mobile: installApp", args);
getDriver().executeScript("mobile: activateApp", args);

**Comment by KazuCocoa** [comment]
Do you say https://appium.github.io/appium-xcuitest-driver/latest/reference/execute-methods/#mobile-launchapp ?
The process argument should be only when the process starts.

**Reply by deepakarorawins** [reply]
Yes, I am talking about that only. However not restricted with it.

E.g. We start our execution with processArgument (--qamode) which gives us some extra details on app screens, so when driver initialized I see those qamode details. However during the my suite execution, I need to terminate/activate/launch the app which make our app loose the extra details which we got with qamode processArgument.


In other words, we want to keep processArguments during the app management

Could you please suggest any other solution if any to my requirement?

**Reply by KazuCocoa** [reply]
> keep processArguments during the app management

I don't understand this point. Do you mean you'd like to give processArguments, which can apply only for app launching time to app terminate as well, while nothing can do for the app?

The `mobile:launchApp` calls https://developer.apple.com/documentation/xctest/xcuiapplication/1500467-launch in the endpoint, so it can give launchArguments, but others like activate and terminate are https://developer.apple.com/documentation/xctest/xcuiapplication/2873317-activate and https://developer.apple.com/documentation/xctest/xcuiapplication/1500637-terminate , which have no way to give `processArguments`.

**Reply by deepakarorawins** [reply]
ok, so there is now way to get processArguments with activate/terminate

I'll try if launch = (remove-->install-->activate)


We have processArgument --qamode
when we start our test with --qamode and it works well
in between test we need 

start test with --qamode (passed in desired capability)
test step1 (pass)
test step2 (pass)

test step 3: getDriver().executeScript("mobile: removeApp", args);
test step4: getDriver().executeScript("mobile: installApp", args);
test step5: getDriver().executeScript("mobile: activateApp", args);

**here we lost the --qamode functionality and our test will fail**

test step6 (fail)
test step7 (fail)
end test

**Reply by KazuCocoa** [reply]
Then, the activateApp should be the launchApp. Only launchApp can pass processArguments. Activate app could keep the process argument only when the process is running launched with the process arguments.

**Reply by deepakarorawins** [reply]
ok, thank you very much for this suggestion, this worked for me :)

**String[] arguments = {"--qamode"};**

Map<String, Object> args = new HashMap<>();
args.put("bundleId", bundleId);
args.put("app", appPath);
args.put("arguments", arguments);
getDriver().executeScript("mobile: removeApp", args);
getDriver().executeScript("mobile: installApp", args);
getDriver().executeScript("mobile: launchApp", args);

---------------------------------------
driver.launchApp(); is depreciated
driver.executeScript("mobile: launchApp", args); is not depreciated, right?
